iceoryx_posh/test/moduletests/test_popo_used_chunk_list.cpp

```diff
@@ -0,0 +1,300 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "iceoryx_posh/internal/popo/used_chunk_list.hpp"
+
+#include "iceoryx_posh/internal/mepoo/memory_manager.hpp"
+#include "iceoryx_posh/mepoo/mepoo_config.hpp"
+#include "iceoryx_utils/internal/posix_wrapper/shared_memory_object/allocator.hpp"
+
+#include "test.hpp"
+
+namespace
+{
+using namespace ::testing;
+using namespace iox::mepoo;
+using namespace iox::popo;
+
+class UsedChunkList_test : public Test
+{
+  public:
+    void SetUp() override
+    {
+        static constexpr uint32_t NUM_CHUNKS_IN_POOL = 100U;
+        static constexpr uint32_t CHUNK_SIZE = 128U;
+        MePooConfig mempoolconf;
+        mempoolconf.addMemPool({CHUNK_SIZE, NUM_CHUNKS_IN_POOL});
+
+        iox::posix::Allocator memoryAllocator{m_memory.get(), MEMORY_SIZE};
+        memoryManager.configureMemoryManager(mempoolconf, &memoryAllocator, &memoryAllocator);
+    };
+
+    void TearDown() override{};
+
+    SharedChunk getChunkFromMemoryManager()
+    {
+        constexpr uint32_t PAYLOAD_SIZE{32U};
+        return memoryManager.getChunk(PAYLOAD_SIZE,
+                                      iox::CHUNK_DEFAULT_PAYLOAD_ALIGNMENT,
+                                      iox::CHUNK_NO_CUSTOM_HEADER_SIZE,
+                                      iox::CHUNK_NO_CUSTOM_HEADER_ALIGNMENT);
+    }
+
+    void createMultipleChunks(uint32_t numberOfChunks, std::function<void(SharedChunk&&)> testHook)
+    {
+        ASSERT_TRUE(testHook);
+        for (uint32_t i = 0; i < numberOfChunks; ++i)
+        {
+            testHook(getChunkFromMemoryManager());
+        }
+    }
+
+    void checkIfEmpty()
+    {
+        SCOPED_TRACE(std::string("Empty check"));
+        for (uint32_t i = 0; i < USED_CHUNK_LIST_CAPACITY; ++i)
+        {
+            EXPECT_TRUE(sut.insert(getChunkFromMemoryManager()));
+        }
+    }
+
+    MemoryManager memoryManager;
+
+    static constexpr uint32_t USED_CHUNK_LIST_CAPACITY{10U};
+    UsedChunkList<USED_CHUNK_LIST_CAPACITY> sut;
+
+  private:
+    static constexpr size_t MEGABYTE = 1U << 20U;
+    static constexpr size_t MEMORY_SIZE = 4U * MEGABYTE;
+    std::unique_ptr<char[]> m_memory{new char[MEMORY_SIZE]};
+};
+
+TEST_F(UsedChunkList_test, OneChunkCanBeAdded)
+{
+    EXPECT_TRUE(sut.insert(getChunkFromMemoryManager()));
+}
+
+TEST_F(UsedChunkList_test, AddSameChunkTwiceWorks)
+{
+    auto chunk = getChunkFromMemoryManager();
+    sut.insert(chunk);
+
+    EXPECT_TRUE(sut.insert(chunk));
+}
```

Comment:
This checks the current behavior. If we want a `Set` like behavior, this needs to be done in a follow up PR.

Reply:
I would opt against a set considering how it is used, since it would imply an error in the memory manager I think (it would have to provide a used chunk again). This adds unnecessary complexity and a performance loss for no gain. We must ensure that memory management works correctly.

As a side note: consider the pair malloc and free. Except for very simple sanity checks,free does not check for double free errors for good reason: the cost would simply be prohibitively high and there is no need if used correctly.

Reply:
it's also no problem to insert the same data multiple times, one just hast to remove it as often as it was inserted

Reply:
Yes, but I just should not happen the way we use it and I do not like guarding against at significant costs (complexity, performance).